### ISSUE TITLE
`./build_openroad.sh` fails because "git -C" is not supported inside Docker

### ISSUE THREAD

**Issue by oharboe** [The-OpenROAD-Project/OpenROAD-flow-scripts#1467]
### Subject

Docker build failure

### Describe the bug

NOTE! "git -C" works locally, but fails inside the Docker image because the Docker image is running an old version of git, whereas git locally on Ubuntu 23.04 is sufficiently new that "git -C" is supported.

```
$ git status
HEAD detached at origin/master
nothing to commit, working tree clean
$ ./build_openroad.sh 
[ 89%] Building techlibs/sf2/synth_sf2.o
[ 99%] Building yosys-config
[ 99%] Building abc/abc-bafd2a7
Unknown option: -C
usage: git [--version] [--help] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p|--paginate|--no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]
ERROR: ABC contains local modifications! Set ABCREV=default in Yosys Makefile!
make: *** [Makefile:745: abc/abc-bafd2a7] Error 1
make: *** Waiting for unfinished jobs....
[ 99%] Building passes/techmap/filterlib.o
passes/techmap/abc.cc: In member function 'virtual void {anonymous}::AbcPass::execute(std::vector<std::basic_string<char> >, Yosys::RTLIL::Design*)':
passes/techmap/abc.cc:1825:15: warning: 'g_argidx' may be used uninitialized in this function [-Wmaybe-uninitialized]
      cmd_error(args, g_argidx, stringf("Unsupported gate type: %s", g.c_str()));
      ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make: Leaving directory '/OpenROAD-flow-scripts/tools/yosys'
The command '/bin/sh -c ./build_openroad.sh --no_init --local --threads ${numThreads}' returned a non-zero code: 2
```

### Expected Behavior

It should build

### Environment

```shell
[WARNING] Your current OpenROAD version is outdated.
It is recommened to pull the latest changes.
If problem persists, file a github issue with the re-producible test case.
kernel: Linux 6.2.0-32-generic
os: Ubuntu 23.04 (Lunar Lobster)
cmake version 3.25.1
-- The CXX compiler identification is GNU 12.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- OpenROAD version: v2.0-10111-gb8c71cc9a
-- System name: Linux
-- Compiler: GNU 12.3.0
-- Build type: RELEASE
-- Install prefix: /usr/local
-- C++ Standard: 17
-- C++ Standard Required: ON
-- C++ Extensions: OFF
-- The C compiler identification is GNU 12.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Python: /usr/bin/python3 (found version "3.11.4") found components: Interpreter 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Performing Test C_COMPILER_SUPPORTS__-Wall
-- Performing Test C_COMPILER_SUPPORTS__-Wall - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wall
-- Performing Test CXX_COMPILER_SUPPORTS__-Wall - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-array-bounds
-- Performing Test C_COMPILER_SUPPORTS__-Wno-array-bounds - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-array-bounds
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-array-bounds - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-nonnull
-- Performing Test C_COMPILER_SUPPORTS__-Wno-nonnull - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-nonnull
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-nonnull - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-maybe-uninitialized
-- Performing Test C_COMPILER_SUPPORTS__-Wno-maybe-uninitialized - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-maybe-uninitialized
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-maybe-uninitialized - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format-overflow
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format-overflow - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format-overflow
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format-overflow - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-variable
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-variable - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-variable
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-variable - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-function
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-function - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-function
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-function - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-write-strings
-- Performing Test C_COMPILER_SUPPORTS__-Wno-write-strings - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-write-strings
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-write-strings - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-sign-compare
-- Performing Test C_COMPILER_SUPPORTS__-Wno-sign-compare - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-sign-compare
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-sign-compare - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-deprecated
-- Performing Test C_COMPILER_SUPPORTS__-Wno-deprecated - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-deprecated
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-deprecated - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-c++11-narrowing
-- Performing Test C_COMPILER_SUPPORTS__-Wno-c++11-narrowing - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-c++11-narrowing
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-c++11-narrowing - Failed
-- Performing Test C_COMPILER_SUPPORTS__-Wno-register
-- Performing Test C_COMPILER_SUPPORTS__-Wno-register - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-register
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-register - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format - SucCMake Warning at src/CMakeLists.txt:233 (message):
  spdlog: SPDLOG_FMT_EXTERNAL=ON


CMake Error at src/stt/CMakeLists.txt:38 (find_package):
  Could not find a package configuration file provided by "LEMON" with any of
  the following names:

    LEMONConfig.cmake
    lemon-config.cmake
    lemonConfig.cmake
    lemon-config.cmake

  Add the installation prefix of "LEMON" to CMAKE_PREFIX_PATH or set
  "LEMON_DIR" to a directory containing one of the above files.  If "LEMON"
  provides a separate development package or SDK, be sure it has been
  installed.


cess
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal
-- Performing Test C_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal - Failed
-- Performing Test C_COMPILER_SUPPORTS__-fpermissive
-- Performing Test C_COMPILER_SUPPORTS__-fpermissive - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-fpermissive
-- Performing Test CXX_COMPILER_SUPPORTS__-fpermissive - Success
-- Performing Test C_COMPILER_SUPPORTS__-x
-- Performing Test C_COMPILER_SUPPORTS__-x - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-x
-- Performing Test CXX_COMPILER_SUPPORTS__-x - Failed
-- Performing Test C_COMPILER_SUPPORTS__c++
-- Performing Test C_COMPILER_SUPPORTS__c++ - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__c++
-- Performing Test CXX_COMPILER_SUPPORTS__c++ - Failed
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-but-set-variable
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-but-set-variable - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-but-set-variable
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-but-set-variable - Success
-- TCL library: /usr/lib/x86_64-linux-gnu/libtcl.so
-- TCL header: /usr/include/tcl/tcl.h
-- TCL readline library: /usr/lib/x86_64-linux-gnu/libtclreadline.so
-- TCL readline header: /usr/include/x86_64-linux-gnu
-- Found SWIG: /usr/bin/swig4.0 (found suitable version "4.1.0", minimum required is "3.0")  
-- Using SWIG >= 4.1.0 -flatstaticmethod flag for python
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0")  
-- boost: 1.74.0
-- Found Python3: /usr/include/python3.11 (found version "3.11.4") found components: Development Development.Module Development.Embed 
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.13") 
-- spdlog: 1.10.0
-- Found BISON: /usr/bin/bison (found version "3.8.2") 
-- Found Doxygen: /usr/bin/doxygen (found version "1.9.4") found components: doxygen dot 
-- STA version: 2.4.0
-- STA git sha: 7cf916ba205115a06c4531a044ced481f1ff8f12
-- System name: Linux
-- Compiler: GNU 12.3.0
-- Build type: RELEASE
-- Build CXX_FLAGS: -O3 -DNDEBUG
-- Install prefix: /usr/local
-- Found FLEX: /usr/bin/flex (found version "2.6.4") 
-- TCL library: /usr/lib/x86_64-linux-gnu/libtcl.so
-- TCL header: /usr/include/tcl/tcl.h
-- SSTA: 0
-- STA executable: /home/oyvind/docker-ORFS/tools/OpenROAD/src/sta/app/sta
-- Configuring incomplete, errors occurred!
```

### To Reproduce

See above.

### Relevant log output

```shell
$ git status
HEAD detached at b8c71cc9a
You are currently bisecting, started from branch '68113315c'.
  (use "git bisect reset" to get back to the original branch)

nothing to commit, working tree clean
$ etc/Env.sh 
[WARNING] Your current OpenROAD version is outdated.
It is recommened to pull the latest changes.
If problem persists, file a github issue with the re-producible test case.
kernel: Linux 6.2.0-32-generic
os: Ubuntu 23.04 (Lunar Lobster)
cmake version 3.25.1
^C
```

**Comment by oharboe** [comment]
Hmmm... running " rm -rf tools/yosys/abc" fixed the issue...  Why would there be files in "tools/yosys/abc"??

**Comment by rovinski** [comment]
That's where the Yosys Makefile builds abc.

**Comment by oharboe** [comment]
> That's where the Yosys Makefile builds abc.

I see. So somehow that folder got poisoned and deleting it fixed the "./build_openroad.sh" docker build command.

Strangely, "git clean -fdx && git submodule foreach git clean -fdx ." didn't clean out that folder...

**Comment by rovinski** [comment]
Yeah, I've run into that struggle before. I don't know why it doesn't get cleaned.

**Comment by oharboe** [comment]
> Yeah, I've run into that struggle before. I don't know why it doesn't get cleaned.

`git submodule foreach` does enter the tools/yosys folder:

```
$ git submodule foreach pwd
Entering 'tools/LSOracle'
/home/oyvind/docker-ORFS/tools/LSOracle
Entering 'tools/OpenROAD'
/home/oyvind/docker-ORFS/tools/OpenROAD
Entering 'tools/yosys'
/home/oyvind/docker-ORFS/tools/yosys
```

**Comment by oharboe** [comment]
```
$ ./build_openroad.sh --latest
[deleted, abc was also built here]
-- Installing: /OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
-- Set runtime path of "/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad" to "/OpenROAD-flow-scripts/dependencies/lib:/OpenROAD-flow-scripts/dependencies/lib64:/opt/or-tools/lib64:/opt/or-tools/lib"
Removing intermediate container bd48610af1dd
 ---> ac4d3762e1ab
Successfully built ac4d3762e1ab
Successfully tagged openroad/flow-centos7-builder:latest
$ ls tools/yosys/abc
ls: cannot access 'tools/yosys/abc': No such file or directory
```

So tools/yosys/abc is not needed to build the docker image...

**Comment by oharboe** [comment]
And if I try to create a file in tools/yosys/abc/def, it is cleaned up...

```
$ mkdir tools/yosys/abc
$ touch tools/yosys/abc/def
$ git submodule foreach git clean -fdx .
Entering 'tools/LSOracle'
Entering 'tools/OpenROAD'
Entering 'tools/yosys'
Removing abc/
$ ls tools/yosys/abc
ls: cannot access 'tools/yosys/abc': No such file or directory
```

**Comment by oharboe** [comment]
So here is a sequence that creates `tools/yosys/abc` that is not cleaned out, after which `./build_openroad.sh` fails:

```
$ ./build_openroad.sh --local
$ git submodule foreach git clean -fdx .
$ ls tools/yosys/abc/
abc-bafd2a7  abc.rc        CMakeLists.txt  i10.aig   readmeaig
abcexe.dsp   abcspace.dsw  copyright.txt   lib       README.md
abclib.dsp   arch_flags.c  depends.sh      Makefile  src
```

**Comment by oharboe** [comment]
Something copies abc into tools/yosys/abc and, I think, because tools/yosys/abc/.git exists, then the tools/yosys/abc folder is not cleaned out by `git clean -fdx .`

```
$ git submodule
 cf0a0eb29523cde8664e9313f31d7f7406d35d31 tools/LSOracle (v0.3.0-rc1-54-gcf0a0eb2)
 b8c71cc9a5bdcc182f98e06e2e473241eef5ecf5 tools/OpenROAD (v2.0-10111-gb8c71cc9a)
 bc027b2cae9a85b887684930705762fac720b529 tools/yosys (yosys-0.13-16-gbc027b2ca)
$ cd tools/yosys
$ git submodule
$ ls abc/.git
branches  description  HEAD   index  logs     packed-refs
config    FETCH_HEAD   hooks  info   objects  refs
```

**Comment by oharboe** [comment]
@maliberty @vvbandeira I now understand the series of unfortunate events that leads to a mysterious failure when trying to build the Docker image:

1. Run `./build_openroad.sh --local`. When Yosys is built, it clones tools/yosys/abc
2. Now run `./build_openroad.sh` to build the docker image, this fails because yosys uses `git -C`, which is not supported by the git version in the docker image, to check if tools/yosys/abc is the right version.

Proposed solution:

1. upgrade git in the docker image so that it supports `git -C`.
2. Make sure only *source code* is copied from the local folder into Docker images. If binary files sneak in from the local build to the Docker images, then you can get error messages when invoking yosys that `libreadline.so` is not available and other mysterious error messages.

**Comment by oharboe** [comment]
Fixed, the docker image is now Ubuntu 22.04, CentOS7 is retired.
